utils/staking-miner/src/rpc.rs

```diff
@@ -0,0 +1,147 @@
+use sp_core::{storage::StorageKey, Bytes};
+use sp_version::RuntimeVersion;
+use std::{future::Future, time::Duration};
+
+#[derive(frame_support::DebugNoBound, thiserror::Error)]
+pub(crate) enum RpcHelperError {
+	JsonRpsee(#[from] jsonrpsee::core::Error),
+	Codec(#[from] codec::Error),
+}
+
+impl std::fmt::Display for RpcHelperError {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		<RpcHelperError as std::fmt::Debug>::fmt(self, f)
+	}
+}
+
+#[rpc(client)]
+pub trait RpcApi {
+	/// Fetch system name
+	#[method(name = "system_chain")]
+	async fn system_chain(&self) -> RpcResult<String>;
+
+	/// Fetch a storage key
+	#[method(name = "state_getStorage")]
+	async fn storage(&self, key: &StorageKey, hash: Option<&Hash>) -> RpcResult<Option<Bytes>>;
+
+	/// Fetch the runtime version
+	#[method(name = "state_getRuntimeVersion")]
+	async fn runtime_version(&self, at: Option<&Hash>) -> RpcResult<RuntimeVersion>;
+
+	#[method(name = "payment_queryInfo")]
+	async fn payment_query_info(
+		&self,
+		encoded_xt: &Bytes,
+		at: Option<&Hash>,
+	) -> RpcResult<RuntimeDispatchInfo<Balance>>;
+
+	/// Dry run an extrinsic at a given block. Return SCALE encoded ApplyExtrinsicResult.
+	#[method(name = "system_dryRun", aliases = ["system_dryRunAt"])]
+	async fn dry_run(&self, extrinsic: &Bytes, at: Option<&Hash>) -> RpcResult<Bytes>;
+
+	/// Submit an extrinsic to watch.
+	///
+	/// See [`TransactionStatus`](sc_transaction_pool_api::TransactionStatus) for details on
+	/// transaction life cycle.
+	//
+	// TODO: https://github.com/paritytech/jsonrpsee/issues/698
+	#[subscription(
+		name = "author_submitAndWatchExtrinsic" => "author_extrinsicUpdate",
+		unsubscribe_aliases = ["author_unwatchExtrinsic"],
+		item = TransactionStatus<Hash, Hash>,
+	)]
+	fn watch_extrinsic(&self, bytes: &Bytes) -> RpcResult<()>;
+
+	/// New head subscription.
+	#[subscription(
+		name = "chain_subscribeNewHeads" => "newHead",
+		item = Header
+	)]
+	fn subscribe_new_heads(&self) -> RpcResult<()>;
+
+	/// Finalized head subscription.
+	#[subscription(
+		name = "chain_subscribeFinalizedHeads" => "chain_finalizedHead",
+		item = Header
+	)]
+	fn subscribe_finalized_heads(&self) -> RpcResult<()>;
+}
+
+/// Wraps a shared websocket JSON-RPC client that can be cloned.
+#[derive(Clone, Debug)]
+pub(crate) struct SharedRpcClient(Arc<WsClient>);
+
+impl Deref for SharedRpcClient {
+	type Target = WsClient;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl SharedRpcClient {
+	/// Consume and extract the inner client.
+	pub fn into_inner(self) -> Arc<WsClient> {
+		self.0
+	}
+
+	/// Create a new shared RPC client.
+	pub(crate) async fn new(uri: &str) -> Result<Self, RpcError> {
+		let client = WsClientBuilder::default()
+			.connection_timeout(Duration::from_secs(20))
+			.max_request_body_size(u32::MAX)
+			.request_timeout(Duration::from_secs(60))
+			.build(uri)
+			.await?;
+		Ok(Self(Arc::new(client)))
+	}
+
+	/// Get the storage item.
```

Comment:
I would expect this doc to explain how this is different than `self.storage`. In the same spirit, I would name it `decode_storage` or `get_storage_decoded`.